agent/src/main/java/com/teamscale/jacoco/agent/store/upload/UploadStoreBase.java

```diff
@@ -94,38 +92,38 @@ protected boolean tryUpload(String xml) {
 		} catch (IOException e) {
 			logger.error("Failed to upload coverage to {}. Probably a network problem", uploadUrl, e);
 			return false;
-		} catch (UploadStoreException e) {
+		} catch (UploaderException e) {
 			logger.error("Failed to upload coverage to {}. The configuration is probably incorrect", uploadUrl, e);
 			return false;
 		}
 	}
 
 	/**
-	 * Creates the zip file to upload which includes the given coverage XML and all
-	 * {@link #additionalMetaDataFiles}.
+	 * Creates the zip file to upload which includes the given coverage XML and all {@link #additionalMetaDataFiles}.
 	 */
-	private byte[] createZipFile(String xml) throws IOException {
-		try (ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream()) {
-			try (ZipOutputStream zipOutputStream = new ZipOutputStream(byteArrayOutputStream);
-				 OutputStreamWriter writer = new OutputStreamWriter(zipOutputStream, "UTF-8")) {
-				fillZipFile(zipOutputStream, writer, xml);
+	private File createZipFile(File coverageFile) throws IOException {
+		File zipFile = createZipFileObject(coverageFile);
+		try (FileOutputStream fileOutputStream = new FileOutputStream(zipFile)) {
+			try (ZipOutputStream zipOutputStream = new ZipOutputStream(fileOutputStream)) {
+				fillZipFile(zipOutputStream, coverageFile);
 			}
-			return byteArrayOutputStream.toByteArray();
+			return zipFile;
 		}
 	}
 
+	private File createZipFileObject(File coverageFile) {
+		Path parentFolder = coverageFile.getParentFile().toPath();
+		String filename = coverageFile.getName().substring(0, coverageFile.getName().length() - 4) + ".zip";
+		return new File(parentFolder.resolve(filename).toString());
+	}
+
 	/**
-	 * Fills the upload zip file with the given coverage XML and all
-	 * {@link #additionalMetaDataFiles}.
+	 * Fills the upload zip file with the given coverage XML and all {@link #additionalMetaDataFiles}.
 	 */
-	private void fillZipFile(ZipOutputStream zipOutputStream, OutputStreamWriter writer, String xml)
+	private void fillZipFile(ZipOutputStream zipOutputStream, File coverageFile)
 			throws IOException {
 		zipOutputStream.putNextEntry(new ZipEntry("coverage.xml"));
-		writer.write(xml);
-
-		// We flush the writer, but don't close it here, because closing the writer
-		// would also close the zipOutputStream, making further writes impossible.
-		writer.flush();
+		zipOutputStream.write(FileSystemUtils.readFileBinary(coverageFile));
 
 		for (Path additionalFile : additionalMetaDataFiles) {
 			zipOutputStream.putNextEntry(new ZipEntry(additionalFile.getFileName().toString()));
```

Comment:
the next line also loads arbitrary files into memory that might be huge. we should also stream this instead of caching them in memory